File: 2191-sort-the-jumbled-numbers/2191-sort-the-jumbled-numbers.cpp
class Solution {
public:
    auto switcharoo(std::vector<int> mapping, int num) -> int {
        std::string str = std::to_string(num);
        int res = 0,
            pow10 = 1;
        
        for (int i = str.size() - 1; i >= 0; --i) {
            // convert according to mapping array with order of magnitude
            int n = mapping[str[i] - '0'] * pow10;
            // add to result;
            res += n;
            // increase order of magnitude for next iteration
            pow10 *= 10;
        }
        
        return res;
    }
    
    vector<int> sortJumbled(vector<int>& mapping, vector<int>& nums) {
        std::map<int, std::vector<int>> map;
        
        for (auto &n:nums) {
            map[switcharoo(mapping, n)].push_back(n);
        }
        
        std::vector<int> res;
        for (auto &[map,vec]:map) {
            res.insert(res.end(), vec.begin(), vec.end());
        }
        
        return res;
    }
};

File: 2191-sort-the-jumbled-numbers/2191-sort-the-jumbled-numbers.rs
impl Solution {
    pub fn sort_jumbled(mapping: Vec<i32>, nums: Vec<i32>) -> Vec<i32> {
        let mut vec_p: Vec<(i32, usize)> = nums.iter()
            .enumerate()
            .map(|(i, num)| (Self::switcharoo(&mapping, num), i))
            .collect();
        
        vec_p.sort();
    
        vec_p.into_iter()
            .map(|(mapped_num, original_idx)| nums[original_idx])
            .collect()
    }
    
    fn switcharoo(mapping: &[i32], num: &i32) -> i32 {
        if *num < 10 {
            mapping[*num as usize]
        } else {
            let mut num_m = num.clone();
            let (mut res, mut pow10) = (0, 1);
            
            while num_m > 0 {
                res += mapping[(num_m % 10) as usize] * pow10;
                pow10 *= 10;
                num_m /= 10;
            }
            
            res
        }
    }
}
